docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+  Kind string `json:"kind"`
+ 
+  // MatchResources selects templates based on where they are referenced.
+  MatchResources PatchMatchResources `json:"matchResources"`
+}
+ 
+// PatchMatchResources selects templates based on where they are referenced.
+// Note: At least one of the fields must be set.
+// Note: The results of selection based on the individual fields are ORed.
+type PatchMatchResources struct {
+  // ControlPlane selects templates referenced in .spec.ControlPlane.
+  // Note: this will match the controlPlane and also the controlPlane 
+  // machineInfrastructure (depending on the kind and apiVersion).
+  // +optional
+  ControlPlane *bool `json:"controlPlane,omitempty"`
+ 
+  // InfrastructureCluster selects templates referenced in .spec.infrastructure.
+  // +optional
+  InfrastructureCluster *bool `json:"infrastructureCluster,omitempty"`
+ 
+  // MachineDeploymentClass selects templates referenced in specific MachineDeploymentClasses in
+  // .spec.workers.machineDeployments.
+  // +optional
+  MachineDeploymentClass *PatchMatchMachineDeploymentClass `json:"machineDeploymentClass,omitempty"`
+}
+ 
+// PatchMatchMachineDeploymentClass selects templates referenced
+// in specific MachineDeploymentClasses in .spec.workers.machineDeployments.
+type PatchMatchMachineDeploymentClass struct {
+  // Names selects templates by class names.
+  Names []string `json:"names"`
+}
+```
+
+**JSONPatchDefinition**
+
+```golang
+// JSONPatchDefinition defines a JSON patch.
+type JSONPatchDefinition struct {
+  // Op defines the operation of the patch.
+  // Note: Only `add`, `replace` and `remove` are supported.
+  Op string `json:"op"`
+ 
+  // Path defines the path of the patch.
+  // Note: Only the spec of a template can be patched, thus the path has to start with /spec/.
+  // Note: For now the only allowed array modifications are `append` and `prepend`, i.e.:
+  // * for op: `add`: only index 0 (prepend) and - (append) are allowed
+  // * for op: `replace` or `remove`: no indexes are allowed
+  Path string `json:"path"`
+ 
+  // Value defines the value of the patch.
+  // Note: Either Value or ValueFrom is required for add and replace
+  // operations. Only one of them is allowed to be set at the same time.
+  // +optional
+  Value *apiextensionsv1.JSON `json:"value,omitempty"`
+ 
+  // ValueFrom defines the value of the patch.
+  // Note: Either Value or ValueFrom is required for add and replace
+  // operations. Only one of them is allowed to be set at the same time.
+  // +optional
+  ValueFrom *JSONPatchDefinitionValue `json:"valueFrom,omitempty"`
+}
+ 
+// JSONPatchDefinitionValue defines the value of a patch.
+// Note: Only one of the fields is allowed to be set at the same time.
+type JSONPatchDefinitionValue struct {
+  // Variable is the variable to be used as value.
+  // Variable can be one of the variables defined in .spec.variables or a builtin variable.
+  // Note: Builtin variables are described in the CAPI book.
```

Comment:
Is this useful context? This isn't in the book right now, right? Won't this and a lot more end up in the book as documentation eventually?

Reply:
I can drop it, but I thought otherwise I would get comments where builtin variables are defined. (as they are not defined here in the struct section of the proposal)

Also eventually we seem to try to keep the proposal structs more or less sync with the ones we have eventually in the code. So I though it's a good idea to add it.

Reply:
I think the best solution is either to describe them here or to leave them undefined - this way splits the difference in a way that doesn't seem to add anything to the proposal.

Reply:
Okay, I'll drop it